Finished the initial transform script.

GabrielNsiah opened the pull request:
The initial transform script is working from start to finish and exports the data of the csv, however the list of genres includes location, we will need a sort of filter to remove locations. For ticket #5.

ebradley12 commented on pipeline/extract.py:
```diff
@@ -30,11 +30,13 @@ def get_sales_information() -> dict:
 
-def main_extract():
+def main_extract() -> dict:
     """Runs all the functions necessary for
     extraction in one go."""
```
Maybe a better docstring would be "Main function to extract data from the Bandcamp API and returns a dictionary containing raw data from the API."

bentrzcinski replied:
```suggestion
    """
    Runs all the functions necessary for
    extraction in one go.
    """
```

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def get_genres_from_url(artist_url: str) -> list[str]:
+    """Gets the lists of genres from a given artist url.
+        Returns an empty list if none can be found."""
```
Could you add a try, except block here as currently, if the API or web scraping fails, the entire pipeline could break. Add exception handling to ensure robustness.

bentrzcinski replied:
```suggestion
    """
    Gets the lists of genres from a given artist url.
    Returns an empty list if none can be found.
    """
```

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+    logging.info("Failed to connect to %s, Status Code: %s",
+                 full_url, response.status_code)
+    return []
```
def get_genres_from_url(artist_url: str) -> List[str]:
    try:
        full_url = convert_to_full_url(artist_url)
        response = requests.get(full_url, timeout=1000)
        response.raise_for_status()  # Raise HTTPError for bad responses

        logging.info("Getting genre tags for %s", full_url)
        soup = BeautifulSoup(response.text, 'html.parser')
        tags = soup.find_all('a', class_='tag')
        return [tag.text for tag in tags] if tags else []

    except requests.RequestException as e:
        logging.error("Failed to connect to %s: %s", artist_url, str(e))
        return []

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+    return sales_info
+
+
+def get_genres_from_url(artist_url: str) -> list[str]:
```
def get_genres_from_url(artist_url: str) -> List[str]:
    try:
        full_url = convert_to_full_url(artist_url)
        response = requests.get(full_url, timeout=1000)
        response.raise_for_status()  # Raise HTTPError for bad responses

        logging.info("Getting genre tags for %s", full_url)
        soup = BeautifulSoup(response.text, 'html.parser')
        tags = soup.find_all('a', class_='tag')
        return [tag.text for tag in tags] if tags else []

    except requests.RequestException as e:
        logging.error("Failed to connect to %s: %s", artist_url, str(e))
        return []

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+    provided in the url, returns an empty string
+    if it cannot be found"""
+
+    full_url = convert_to_full_url(artist_url)
```
Refactor common code - get_genres_from_url and get_release_date both call requests.get and parse the response with BeautifulSoup.

def fetch_webpage(url: str) -> BeautifulSoup:
    try:
        full_url = convert_to_full_url(url)
        response = requests.get(full_url, timeout=1000)
        response.raise_for_status()
        return BeautifulSoup(response.text, 'html.parser')
    except requests.RequestException as e:
        logging.error("Failed to fetch webpage %s: %s", url, str(e))
        return BeautifulSoup("", 'html.parser')

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+    if response.status_code == 200:
+        logging.info("Getting release date for for %s", full_url)
+        soup = BeautifulSoup(response.text, 'html.parser')
```
soup = fetch_webpage(artist_url)

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def create_dataframe(sales_info: list[dict]) -> pd.DataFrame:
+    """Returns a dataframe based on received sales information"""
```
Could the docstring be "Converts sales information into a Pandas DataFrame and saves it to a CSV file."

bentrzcinski replied:
```suggestion
    """Returns a dataframe based on received sales information."""
```

emily-curtis commented on pipeline/extract.py:
```diff
@@ -30,11 +30,13 @@ def get_sales_information() -> dict:
+def main_extract() -> dict:
     """Runs all the functions necessary for
     extraction in one go."""
     config_log()
```
also maybe use the following format for 2 line docstrings:
"""
...
"""

bentrzcinski commented on pipeline/extract.py:
```diff
@@ -30,11 +30,13 @@ def get_sales_information() -> dict:
     extraction in one go."""
     config_log()
-    get_sales_information()
+    sales_data = get_sales_information()
+
+    return sales_data
```
```suggestion
    return get_sales_information()
```

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def config_log() -> None:
+    """Terminal logs configuration"""
```
```suggestion
    """Terminal logs configuration."""
```

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+def get_sale_information(sales_dict: dict) -> list[dict]:
+    """Returns a list of dictionaries containing the
+    sales information"""
```
This function handles both extracting genres and release dates. Could you split into smaller functions?

GabrielNsiah replied:
I believe having the nested function calls of `get_release_date` and ` get_genres_from_url` are fine, as if I were to split them out as their own standalone functions, it may be harder to follow the flow of information.

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def convert_from_unix_to_datetime(unix: str) -> datetime:
+    """Converts a unix timestamp to a regularly
+     formatted date"""
```
```suggestion
    """
    Converts a unix timestamp to a regularly
     formatted date.
     """
```

emily-curtis commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+def convert_from_unix_to_datetime(unix: str) -> datetime:
+    """Converts a unix timestamp to a regularly
+     formatted date"""
+
```
Again try to format the docstring like I commented above maybe?

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+    """Converts a unix timestamp to a regularly
+     formatted date"""
+
+    date = datetime.fromtimestamp(unix)
+
+    return date
```
```suggestion
    return datetime.fromtimestamp(unix)
```

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def convert_to_full_url(url: str) -> str:
+    """Checks if a url is valid and returns
+    the full url if it is not"""
```
```suggestion
    """
    Checks if a url is valid and returns
    the full url if it is not.
    """
```

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+    if url.startswith("http"):
+        return url
+
+    return "https:" + url
```
Does this account for the '//' at the start of a URL? (i.e https://google.com)

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+
+def get_release_date(artist_url: str) -> str:
+    """Retrieves the release of a song or album
+    provided in the url, returns an empty string
+    if it cannot be found"""
```
```suggestion
    """
    Retrieves the release of a song or album
    provided in the url, returns an empty string
    if it cannot be found.
    """
```

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+    return []
+
+
+def get_release_date(artist_url: str) -> str:
```
```suggestion
def get_release_date_from_url(artist_url: str) -> str:
```
Keep the naming consistent with other functions and how they behave?

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+    """Returns a dataframe based on received sales information"""
+    sales_df = pd.json_normalize(sales_info)
+
+    sales_df.to_csv("MUSIC_DATA.csv")
+
+    return sales_df
```
In the long run we will want to handle this as a Pandas DataFrame

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,145 @@
+
+
+if __name__ == "__main__":
+    config_log()
+    sales_data = main_extract()
+    sales_info = get_sale_information(sales_data)
+    create_dataframe(sales_info)
```
Could we put all these in a transform_main() function like with extract?

bentrzcinski reviewed (changes requested):
Nice code, mostly these are just me being picky with doc strings!

GabrielNsiah commented:
The .csv file is being pushed for documentation purposes, and will not be included in the merge.

Changes:
- All docstrings are now formatted with punctuation.
- Some additional functions for data validation and formatting.
- All dates are now in the same 'DD-MM-YYYY' format for the sake of clarity.
- Error handling added to functions.
- Miscellaneous spelling/variable name changes for correctness' sake.

ebradley12 commented on pipeline/MUSIC_DATA.csv:
```diff
@@ -0,0 +1,19 @@
+3,a,A Fenestrated Fantasy,None,silk,France,1.9,"['punk', 'emo', 'emoviolence', 'screamo', 'New York']",01-12-2024,03-12-2024
+4,a,Hidden Treasures of Egypt,None,Skhemty,Spain,3.16,"['ambient', 'ancient egypt', 'dark ambient', 'dark dungeon music', 'dungeon synth', 'egyptian', 'egyptian ambient', 'egyptian music', 'egyptian synth', 'folklore', 'middle eastern', 'orchestral', 'oriental', 'pyramids', 'soundtrack', 'France']",01-12-2024,03-12-2024
+5,t,Thank You,Divine Essence,Svava,United States,1.29,"['a course in miracles', 'christian', 'devotional', 'inspirational', 'Duchesne']",04-03-2019,03-12-2024
+6,a,OCS LIVE AT PERMANENT ROADHOUSE,None,OCS,United Kingdom,7.0,"['folk', 'flok', 'kraut', 'psych', 'Los Angeles']",,03-12-2024
```
I noticed this row is missing a date - could we either remove this row completely or replace with NULL?

ebradley12 commented on pipeline/MUSIC_DATA.csv:
```diff
@@ -0,0 +1,19 @@
+,item_type,item_description,album_title,artist_name,country,amount_paid_usd,genres,release_date,sale_date
+0,t,Johanna in the 80s,None,Sweeney,Australia,6.5,"['adelaide', 'alternative', 'australian', 'electronic pop', 'queerpop', 'synthpop', 'Australia']",02-12-2024,03-12-2024
+1,a,Steve Rothery Band - Live in Frankfurt,None,Steve Rothery,United Kingdom,8.87,"['progressive rock', 'rock', 'Aylesbury']",30-11-2024,03-12-2024
+2,a,UFO 50,None,Eirik Suhrke,United States,10.0,"['chiptune', 'soundtrack', 'video game music', 'jazz', 'synth', 'Norway']",02-10-2024,03-12-2024
```
Seems to be no title for some data - I think we should discuss in next meeting how might be best to handle missing data

ebradley12 commented on pipeline/MUSIC_DATA.csv:
```diff
@@ -0,0 +1,19 @@
+14,a,"Greatest Hits, Vol. 1",None,The Browns,United Kingdom,2.14,"['punk', 'horror punk', 'movies', 'rock', 'Canada']",31-10-1999,03-12-2024
+15,a,Behind The Shadows Lie Madness,None,Mithras,United States,16.48,"['death metal', 'experimental brutal death metal', 'extreme metal', 'metal', 'dark ambient', 'United Kingdom']",16-04-2007,03-12-2024
+16,a,Chiffons d'azur,None,Lamasz & Thme,United Kingdom,0.53,"['ambient', 'electronic', 'ambient', 'ambient electronic', 'chill', 'drone', 'drone ambient', 'field recordings', 'lofi', 'soundscapes', 'tape', 'Paris']",19-11-2024,03-12-2024
+17,a,Apprendre l'Ourdou - Volume 1 & 2,None,Nouredil Marlawi - Langues de l'Inde,France,1.05,"['audiobooks', ""langues de l'inde"", 'ourdou', 'hindi', 'France']",13-11-2024,03-12-2024
```
Nice date formatting

ebradley12 commented on pipeline/transform.py:
```diff
@@ -0,0 +1,211 @@
+    sales_df = pd.json_normalize(sales_info)
+
+    sales_df.to_csv("MUSIC_DATA.csv")
+
```
Could we add the sale year and month to the dataframe just for the trend analysis for the dashboard?

df["sale_year"] = df["sale_date"].dt.year
df["sale_month"] = df["sale_date"].dt.month

ebradley12 reviewed (commented):
Really nice work
Great code

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,211 @@
+    Returns True if it is.
+    """
+
+    if item_type not in ["a", "t"]:
+        return False
+    return True
```
```suggestion
    return item_type in ["a", "t"]
```

bentrzcinski reviewed (changes requested):
Beautiful doc strings!

ebradley12 commented on pipeline/MUSIC_DATA.csv:
```diff
@@ -0,0 +1,21 @@
+t,Heartland,None,The Fulton Collective,Ireland,1.05,"['alternative rock', 'rock', 'Tullamore', 'alternative']",25-11-2024,2024-03-12,2024,3
+t,꧁ঔৣ☬✞ ARCA X ELJORDAN23 ✞☬ঔৣ꧂,None,666succi,United Kingdom,1.0,"['reggaeton', 'bass', 'techno', 'dubstep', 'Valparaíso', 'experimental', 'trap']",01-06-2023,2024-03-12,2024,3
+t,Raw Thrills (Original Mix),Raw Thrills,David Phoenix,Germany,1.5,['Los Angeles'],18-08-2023,2024-03-12,2024,3
+a,Rare Baby,None,KUKII,United Kingdom,1.27,"['mahraganat', 'alternative pop', 'alternative', 'kukii', 'Paris', 'lafawndah']",20-11-2024,2024-03-12,2024,3
```
Can we remove locations from genres using NLP

GabrielNsiah commented:
I've added:
- Filtering for genres (Locations no longer show up in the finished .csv)
- Album names are automatically filled in with the item description if the field is "None" and the item is an album.
- Columns `sales_month` and `sales_year` added for easier access for analysis.
- Some additional functions for cleaning and formatting.
- Fixed a double logging problem from the previous revision.

Once approved I will make another merge, to remove the .csv file.

emily-curtis reviewed (approved):
Looks great to me, thanks Gabriel!

bentrzcinski commented on pipeline/MUSIC_DATA.csv:
```diff
@@ -0,0 +1,70 @@
+item_type,item_description,album_title,artist_name,country,amount_paid_usd,genres,release_date,sale_date,sale_year,sale_month
```
Probably want to add *.csv to .gitignore so we don't clog up the repo with data files

GabrielNsiah replied:
Added.

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,299 @@
+    sales_df["sale_month"] = sales_df["sale_date"].dt.month
+
+
+def replace_blank_album_titles(sales_df: pd.DataFrame) -> None:
```
Don't we want to do this with tracks? My understanding was that tracks commonly had their title in the 'item_description' field.

GabrielNsiah replied:
It was albums that had their name in the `item_description` column, but was lacking it in the `album_title` column.

Tracks always have their name in `item_description` but things like singles don't necessarily have an associated album, so it would be right to have some "None" values for them in the table.

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,299 @@
+
+        event_information = {"item_type": items["item_type"],
+                             "item_description": items["item_description"],
+                             "album_title": items["album_title"],
```
Although this is what the API calls it, maybe we could rename 'album_name' to something like 'release_name' here to better reflect capturing both albums and tracks, and better match with the ERD?

ebradley12 replied:
Great point ^

GabrielNsiah replied:
I've changed:
- `item_type` to `release_type` as we're explicitly dealing with album/track drops.
- `item_description` to `release_name` as it better describes what it is in relation to the table.

Since `album_name` is actually the name of an album that is either being sold as a whole or the album that a track is being sold from, I think it's fine to leave that as is.

bentrzcinski commented on pipeline/transform.py:
```diff
@@ -0,0 +1,299 @@
+    sales_info = get_sale_information(sales_data)
+    sales_df = create_sales_dataframe(sales_info)
+    cleaned_sales_df = clean_sales_dataframe(sales_df)
+    cleaned_sales_df.to_csv("MUSIC_DATA.csv", index=None)
```
Although this is good for testing, probably don't want it in the prod build

GabrielNsiah replied:
If this iteration is approved, I'll remove the .csv output and the file from the push.

bentrzcinski reviewed (changes requested):
Really nice and very comprehensive code, just a few general questions from me really.

ebradley12 reviewed (approved):
Amazing work - thanks Gabriel :)

GabrielNsiah commented:
With this final round of changes I've:

- Added `geonamescache` to `requirements.txt`.
- Added `*.csv` to the `.gitignore`.
- Changed some column names for the sake of clarity.

bentrzcinski reviewed (approved):
All looks good! Thanks for making so many changes.
